Test code to chase down an OSX failure

leculver opened the pull request:
DO NOT MERGE

mikem8361 commented on src/Microsoft.Diagnostics.ExtensionCommands/GCRootCommand.cs:
```diff
@@ -108,6 +109,30 @@ public override void Invoke()
             }
 
             Console.WriteLine($"Found {count:n0} unique roots.");
```
If you use Trace.TraceInformation is will go into the separate *.soslog files and not interfere with the regex's.

leculver replied:
Ah, I was thinking the regex would still work because I only tacked on output to the end.

Good idea though, I'll push a Trace.TraceInformation change after CI finishes processing this run.

leculver commented:
@mikem8361  hmm, interesting.  The regex seems wrong here.  Did I miss an update to the regular expressions, or is this not the failure?

       45 SOSCOMMAND:GCRoot <PREVPOUT>
       46 VERIFY:.*Thread <HEXVAL>:
       47 VERIFY:\s+<HEXVAL>\s+<HEXVAL>\s+GCWhere\.Main\(\)\s+

But the output looks like this:

        00:06.850:           -> 000181536e28     GCWhere 

Code:

                objectOutput.WriteRow("->", obj, obj.Type, extraText);

```

        00:06.753: <END_COMMAND_OUTPUT>
        STDIN: 00:06.753: runcommand sos GCWhere 000181536e28
        00:06.753: (lldb) runcommand sos GCWhere 000181536e28
        00:06.762: Address          Heap   Segment          Generation Allocated           Committed           Reserved           
        00:06.763: 000181536e28     0      000181534000     0          181535000-181538008 181534000-181546000 181546000-191534000
        00:06.763: <END_COMMAND_OUTPUT>
        STDIN: 00:06.763: runcommand sos GCRoot 000181536e28
        00:06.763: (lldb) runcommand sos GCRoot 000181536e28
        00:06.770: Caching GC roots, this may take a while.
        00:06.770: Subsequent runs of this command will be faster.
        00:06.771: 
        00:06.845: Thread e1b2:
        00:06.846:     7ff7be9b7d40 11b2c0b0c
        00:06.846:         rbp-38: 00007ff7be9b7d98
        00:06.850:           -> 000181536e28     GCWhere 
        00:06.850: 
        00:06.850:     7ff7be9b7d40 11b2c0b0c
        00:06.850:         rbp-8: 00007ff7be9b7dc8
        00:06.850:           -> 000181536e28     GCWhere 
        00:06.851: 
        00:06.851: Found 2 unique roots.
        00:06.851: 
        00:06.851: Cached roots in RootCacheService:
        00:06.851: thread:0000e1b2 reg:rbp regoffs:-40 address:7ff7be9b7d90 obj:181536e40 type:System.IO.StringWriter
        00:06.851: thread:0000e1b2 reg:rbp regoffs:-38 address:7ff7be9b7d98 obj:181536e28 type:GCWhere
        00:06.851: thread:0000e1b2 reg:rbp regoffs:-10 address:7ff7be9b7dc0 obj:181536e40 type:System.IO.StringWriter
        00:06.851: thread:0000e1b2 reg:rbp regoffs:-8 address:7ff7be9b7dc8 obj:181536e28 type:GCWhere
        00:06.851: 
        00:06.851: Roots from ClrMD:
        00:06.894: 
        00:06.894: <END_COMMAND_OUTPUT>
    SOSRunner error at /Users/runner/work/1/s/src/SOS/SOS.UnitTests/Scripts/GCTests.script:47
    Excerpt from /Users/runner/work/1/s/src/SOS/SOS.UnitTests/Scripts/GCTests.script:
       45 SOSCOMMAND:GCRoot <PREVPOUT>
       46 VERIFY:.*Thread <HEXVAL>:
       47 VERIFY:\s+<HEXVAL>\s+<HEXVAL>\s+GCWhere\.Main\(\)\s+
       48 
       49 SOSCOMMAND:GCRoot <PREVPOUT>
        00:06.897: <END_COMMAND_ERROR>
        STDIN: 00:06.897: runcommand sos SOSStatus
        00:06.897: (lldb) runcommand sos SOSStatus
        00:06.901: Target OS: OSX Architecture: X64 ProcessId: 18401 (0x47E1)
        00:06.901: #0 .NET Core runtime 8.0.23.12704 at 0000000101544000 size 00001190 index 3209bf50cdab3687a8cb510168958e88
        00:06.901:     Single-file runtime module path: /Users/runner/work/1/s/artifacts/Debuggees/SingleFile/GCWhere/bin/Debug/net8.0/osx-x64/publish/GCWhere
        00:06.901:     DAC: /Users/runner/.dotnet/symbolcache/libmscordaccore.dylib/mach-uuid-3db3f05b7b763af989216365097c9753/libmscordaccore.dylib
        00:06.901: 
        00:06.901: Current symbol store settings:
        00:06.901: -> Directory: /Users/runner/work/1/s/artifacts/Debuggees/SingleFile/GCWhere/bin/Debug/net8.0/osx-x64/publish
        00:06.901: -> Cache: /Users/runner/.dotnet/symbolcache
        00:06.901: -> Server: https://msdl.microsoft.com/download/symbols/ Timeout: 4 RetryCount: 0
        00:06.908: GC memory usage for managed SOS components: 2,247,816 bytes
```

leculver commented:
Oh!  I see, it's missing the stack trace frame, that's odd.  Still digging into it.

leculver commented:
I haven't seen the "Found 0 roots" issue yet.

I did notice that the stack trace is sometimes empty.  This is outside of ClrMD's control, as far as I can tell.  We seem to either fail to create a IXClrDataStackWalk, or the thread context we get back is busted, or some other stackwalking failure.

So far it always seems like the first stack walk succeeds, and subsequent walks may fail (but usually don't).  Since I can't get that under a debugger, I added a cache for stack frames.  I'm trying that as a mitigation while I try to reproduce the other issue.

leculver commented:
Finally saw the failure.  I've added some tracing in ClrMD.  Waiting for the build to make its way through the pipeline.

These failures seem very intermittent.  ClrMD's code is very simple, we just call into the dac API and more or less produce what it shows.  The issue is likely somewhere in either the dac stackwalking api, or within GetThreadContext.  The added tracing should help narrow it down.

leculver commented:
An update here.

We've narrowed this down to a particular test hitting an error HRESULT when attempting to read memory out of the target process:

    Information: 0 : MemoryServiceFromDebuggerServices::ReadMemory: f14ebe88b0 req:8 read:0 hr=8007001e

I was able to reproduce it locally by running the SOS tests and saw it once.  I'm attempting to reproduce it by running _roughly_ the same scenario overnight.  Using ClrMD's dbgeng plugin system I built this plugin:

```C#
    public class CheckRootsCommand : DbgEngCommand
    {
        public CheckRootsCommand(nint pUnknown, bool redirectConsoleOutput = false) : base(pUnknown, redirectConsoleOutput)
        {
        }

        public void Run()
        {
            (int HResult, string Output) result = RunCommandWithOutput("!dso");
            Console.WriteLine($"dso: {result.Output}");
            foreach (string line in result.Output.Split('\n'))
            {
                if (line.Contains("GCWhere"))
                {
                    string[] parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
                    ulong address = ulong.Parse(parts[1], System.Globalization.NumberStyles.HexNumber);

                    result = RunCommandWithOutput($"!gcroot {address:x}");
                    Console.WriteLine($"gcroot: {result.Output}");
                    break;
                }
            }

            if (result.Output.Contains("Found 2 unique roots."))
            {
                Environment.Exit(100);
            }
        }
    }
```

Which minimally replicates the scenario.  I'm going to run it overnight with this and see if it halts:

``` bat
FOR /L %%G IN (1, 1, 99999999) DO (
    C:\Users\leculver\.nuget\packages\cdb-sos\10.0.18362\runtimes\win-x64\native\cdb.exe  -y "D:\git\dotnet\diagnostics\artifacts\bin\GCWhere\Debug\net7.0" -c "g;.load D:\git\dotnet\diagnostics\artifacts\bin\Windows_NT.x64.Debug\sos.dll;.load d:\git\clrmd\artifacts\bin\DbgEngExtension\Debug\net7.0\win-x64\native\DbgEngExtension.dll;!checkroots" -Gsins D:\git\dotnet\diagnostics\.dotnet-test\dotnet.exe --fx-version 7.0.5 D:\git\dotnet\diagnostics\artifacts\bin\GCWhere\Debug\net7.0\GCWhere.dll
    IF NOT ERRORLEVEL 100 (
        echo Error: Unexpected exit code
        GOTO :exit_loop
    )
    echo Iteration: %%G
)
:exit_loop
```

We'll see how it goes.

leculver commented:
(I'm about to go on leave, so I'm being very verbose in this thread in case I have to go.)

No repro after 20,000 iterations.  I think that means the issue has to do with cached data in either the dac or ClrMD since this only seems to reproduce on a live target, and not on the first call to !gcroot.

I added this log message to the only place ClrMD is flushed, and it doesn't show up in the .sslog file:

```C#
            if (_serviceContainer.TryGetCachedService(typeof(ClrRuntime), out object service))
            {
                ((ClrRuntime)service).FlushCachedData();
                Trace.TraceInformation($"Flushing ClrMD");
            }
            else
            {
                Trace.TraceInformation($"Failed to flush ClrMD");
            }
```

I'm trying again with manual flushing before running !gcroot to see if that resolves the issue or not.

mikem8361 commented:
I might see why the ClrRuntime isn't being flushed in Runtime.Flush().   The RuntimeService.Flush is called first (though there isn't any order guarantee) which disposes of all the Runtime instances (instead of flushing) which Runtime.Dispose removes/deregisters the OnFlushEvent handler so Runtime.Flush is never called.

If you followed that great, but it is confusing (and why there might be bugs).  The intent was to cleanup (flush), remove and dispose of Runtime instances on Flush so the RuntimeService re-creates them the next time a runtime service instance is requested. What got lost in one of my big changes was the "cleanup" flush part.  The simple fix is to put a Flush in Runtime.Dispose().

ge ID:
